resolver/bootstrap.go

```diff
@@ -0,0 +1,153 @@
+package resolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/0xERR0R/blocky/config"
+	"github.com/0xERR0R/blocky/log"
+	"github.com/0xERR0R/blocky/model"
+	"github.com/0xERR0R/blocky/redis"
+	"github.com/0xERR0R/blocky/util"
+	"github.com/miekg/dns"
+	"github.com/sirupsen/logrus"
+)
+
+// Bootstrap allows resolving hostnames using the configured bootstrap DNS.
+type Bootstrap struct {
+	resolver Resolver
+}
+
+// NewBoostrap creates and returns a new Boostrap.
+// Internaly, it uses a CachingResolver and an UpstreamResolver.
+func NewBoostrap(cfg *config.Config, redisClient *redis.Client) (b *Bootstrap, err error) {
+	upstream := cfg.BootstrapDNS.Upstream
+
+	var ips []net.IP
+	if upstream.Net == config.NetProtocolTcpUdp {
+		ip := net.ParseIP(upstream.Host)
+		if ip == nil {
+			return nil, fmt.Errorf("bootstrapDns uses %s but is not an IP", upstream.Net)
+		}
+
+		ips = append(ips, ip)
+	} else if cfg.BootstrapDNS.IP == nil {
+		return nil, fmt.Errorf("bootstrapDns.IP is required when upstream uses %s", upstream.Net)
+	} else {
+		ips = append(ips, cfg.BootstrapDNS.IP)
+	}
+
+	chain := Chain(
+		NewCachingResolver(cfg.Caching, redisClient),
+		NewBootstrapedUpstreamResolver(upstream, ips),
+	)
+
+	b = &Bootstrap{
+		resolver: chain,
+	}
+	return
+}
+
+// NewHttpTransport returns a new http.Transport that uses b to resolve hostnames
+func (b *Bootstrap) NewHttpTransport() *http.Transport {
+	dialer := net.Dialer{}
+	logger := log.PrefixedLog("BoostrapDNS")
+
+	return &http.Transport{
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			logger.WithField("network", network).WithField("addr", addr).Tracef("dialing using bootstrap resolver")
+
+			host, port, err := net.SplitHostPort(addr)
+			if err != nil {
+				return nil, err
+			}
+
+			if !strings.Contains("46", network[len(network)-1:]) {
+				// Choose a random IP version to avoid doing two lookups
+				// since we only need a single IP
+				if rand.Intn(1) == 0 {
+					network = network + "4"
+				} else {
+					network = network + "6"
+				}
+			}
+
+			// Resolve the host with the bootstrap DNS
+			var ips []net.IP
+			if strings.HasSuffix(network, "4") {
+				ips, err = b.lookupIP(host, dns.TypeA, logger)
+			} else if strings.HasSuffix(network, "6") {
+				ips, err = b.lookupIP(host, dns.TypeAAAA, logger)
+			}
+			if err != nil {
+				return nil, err
+			}
+
+			if len(ips) == 0 {
+				return nil, fmt.Errorf("boostrap DNS server returned no IPs for %s", host)
+			}
+
+			ip := ips[rand.Intn(len(ips))]
+			addr = net.JoinHostPort(ip.String(), port)
+
+			// Use the standard dialer to connect to the IP
+			return dialer.DialContext(ctx, network, addr)
+		},
+	}
+}
+
+// LookupIP resolves the hostname
+func (b *Bootstrap) LookupIP(hostname string) (ips []net.IP, err error) {
```

Comment:
This function can probably be factorized with `BlockingResolver.queryForFQIdentifierIPs` and made to do the lookups in parallel.
I'll try that if it sounds good to you.